src/crab/ebpf_domain.cpp

```diff
@@ -119,243 +121,644 @@ static std::vector<linear_constraint_t> assume_bit_cst_interval(const NumAbsDoma
         return {};
-    int64_t src_int_value = src_n.value().cast_to_sint64();
 
-    if (!is64) {
-        auto dst_interval = inv.eval_interval(dst_value);
-        std::optional<number_t> dst_lb = dst_interval.lb().number();
-        std::optional<number_t> dst_ub = dst_interval.ub().number();
-        if (!dst_lb || !dst_lb.value().fits_cast_to_int64() || !dst_ub || !dst_ub.value().fits_cast_to_int64())
-            return {};
-        int32_t dst_lb_value = (int32_t)dst_lb.value().cast_to_sint64();
-        int32_t dst_ub_value = (int32_t)dst_ub.value().cast_to_sint64();
-        src_int_value = (int32_t)src_int_value;
-        switch (op) {
-        case Op::EQ:
-            if (dst_lb_value == src_int_value && dst_ub_value == src_int_value)
-                return {linear_constraint_t::TRUE()};
-            if (dst_lb_value > src_int_value || dst_ub_value < src_int_value)
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::NE:
-            if (dst_lb_value == src_int_value && dst_ub_value == src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_lb_value > src_int_value || dst_ub_value < src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        case Op::SGE:
-            if (dst_lb_value >= src_int_value)
-                return {linear_constraint_t::TRUE()};
-            if (dst_ub_value < src_int_value)
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::SLT:
-            if (dst_lb_value >= src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_ub_value < src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        case Op::SLE:
-            if (dst_ub_value <= src_int_value)
-                return {linear_constraint_t::TRUE()};
-            if (dst_lb_value > src_int_value)
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::SGT:
-            if (dst_ub_value <= src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_lb_value > src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        default: throw std::exception();
-        }
-    } else {
-        switch (op) {
-        case Op::EQ: return {dst_value == src_int_value};
-        case Op::NE: return {dst_value != src_int_value};
-        case Op::SGE: return {dst_value >= src_int_value};
-        case Op::SLE: return {dst_value <= src_int_value};
-        case Op::SGT: return {dst_value > src_int_value};
-        case Op::SLT: return {dst_value < src_int_value};
-        default: throw std::exception();
+    // For a 64-bit operation, we can do direct comparisons.
+    if (is64)
+        return {left_svalue >= *right_interval.lb().number(), left_svalue <= *right_interval.ub().number()};
```

Comment:
What happens when the right interval is positive? How does the information propagate to the uvalue? Is this case handled elsewhere?

(this should also add svalue=uvalue to the invariants)